techlibs/quicklogic/synth_quicklogic.cc

```diff
@@ -0,0 +1,162 @@
+        size_t argidx;
+        for (argidx = 1; argidx < args.size(); argidx++)
+        {
+            if (args[argidx] == "-top" && argidx+1 < args.size()) {
+                top_opt = "-top " + args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-edif" && argidx+1 < args.size()) {
+                edif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-blif" && argidx+1 < args.size()) {
+                blif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-flatten") {
+                flatten = true;
+                continue;
+            }
+            break;
+        }
+        extra_args(args, argidx, design);
+
+        if (!design->full_selection())
+            log_cmd_error("This command only operates on fully selected designs!\n");
+
+        log_header(design, "Executing SYNTH_QUICKLOGIC pass.\n");
+        log_push();
+
+        run_script(design, run_from, run_to);
+
+        log_pop();
+    }
+
+    void script() YS_OVERRIDE
+    {
+        if (check_label("begin")) {
+            run("read_verilog -lib +/quicklogic/cells_sim.v");
+            run(stringf("hierarchy -check %s", top_opt.c_str()));
+        }
+
+        if (check_label("prepare")) {
+            run("proc");
+            if (flatten || help_mode)
+                run("flatten", "(with '-flatten')");
+        }
+
+        if (check_label("coarse")) {
+            run("tribuf -logic");
+            run("opt_expr");
+            run("opt_clean");
+            run("check");
+            run("opt");
+            run("peepopt");
+            run("opt_clean");
+            run("techmap");
+            run("abc -lut 1:4");
+            run("opt_clean");
+        }
+
+        if (check_label("map")) {
+            run("techmap -map +/quicklogic/cells_map.v");
+            run("clean");
+            run("check");
+            run("opt_clean -purge");
+        }
+
+        if (check_label("iomap")) {
+            if (help_mode) {
+                run("select -module top");
+            } else {
+                this->currmodule = this->active_design->top_module()->name.str();
+                if (this->currmodule.size() > 0)
+                    run(stringf("select -module %s", this->currmodule.c_str()));
+            }
+            run("clkbufmap -buf $_BUF_ Y:A -inpad ckpad Q:P");
+            run("iopadmap -bits -outpad outpad A:P -inpad inpad Q:P -tinoutpad bipad EN:Q:A:P -ignore ckpad P");
+            if (help_mode) {
+                run("select -clear");
+            } else if (this->currmodule.size() > 0)
+                run("select -clear");
+        }
+
+        if (check_label("finalize")) {
+            run("splitnets -ports -format ()");
+            run("hilomap -hicell logic_1 a -locell logic_0 a -singleton");
+            run("techmap -map +/quicklogic/cells_map.v");
+            run("setundef -zero -params -undriven");
```

Comment:
`setundef` should probably go before `hilomap`, otherwise any zero drivers it creates (for undriven signals) won't be mapped to a constant driver cell.

Reply:
Fixed that, and added a test